crates/connector_proxy/src/libs/image_inspect.rs

```diff
@@ -0,0 +1,88 @@
+use crate::apis::FlowRuntimeProtocol;
+use crate::errors::{Error, Must};
+use clap::ArgEnum;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::BufReader;
+
+// The key of the docker image label that indicates the connector protocol.
+const FLOW_RUNTIME_PROTOCOL_KEY: &str = "FLOW_RUNTIME_PROTOCOL";
+const CONNECTOR_PROTOCOL_KEY: &str = "CONNECTOR_PROTOCOL";
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "PascalCase")]
+pub struct ImageConfig {
+    pub entrypoint: Vec<String>,
+    pub labels: Option<HashMap<String, String>>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "PascalCase")]
+pub struct ImageInspect {
+    pub config: ImageConfig,
+    pub repo_tags: Option<Vec<String>>,
+}
+
+impl ImageInspect {
+    pub fn parse_from_json_file(path: Option<String>) -> Result<Self, Error> {
+        if path.is_none() {}
+        match path {
+            None => {
+                return Err(Error::MissingImageInspectFile);
+            }
+            Some(p) => {
+                let reader = BufReader::new(File::open(p)?);
+                let image_inspects: Vec<ImageInspect> = serde_json::from_reader(reader)?;
+                match image_inspects.len() {
+                    1 => Ok(image_inspects[0].clone()),
+                    _ => Err(Error::InvalidImageInspectFile),
+                }
+            }
+        }
+    }
+
+    pub fn get_entrypoint(&self, default: Vec<String>) -> Vec<String> {
+        match self.config.entrypoint.len() {
+            0 => {
+                tracing::warn!(
+                    "No entry point is specified in the image, using default: {:?}",
+                    default
+                );
+                default
+            }
+            _ => self.config.entrypoint.clone(),
+        }
+    }
+
+    pub fn infer_runtime_protocol(&self) -> FlowRuntimeProtocol {
+        if let Some(ref labels) = self.config.labels {
+            if let Some(value) = labels.get(FLOW_RUNTIME_PROTOCOL_KEY) {
+                return FlowRuntimeProtocol::from_str(&value, false).or_bail();
+            }
+        }
+
+        if let Some(repo_tags) = &self.repo_tags {
+            for tag in repo_tags {
+                if tag.starts_with("ghcr.io/estuary/materialize-") {
```

Comment:
Yes! The intention is that we move to docker labels for all of our published connectors, where those labels carry metadata like whether it's a capture or materialization, and its specific protocol.

We haven't done this yet, though, and this is a short-term hack to keep things working.

Just in case it's useful: you can ask docker to build an image, or locally re-tag an image, to ghcr.io/estuary/materialize-my-thing, even if you have no intention of pushing that up as an image.

Reply:
@jgraettinger I put a TODO here for us to revisit this 👍🏽